src/cli/bin.ts

```diff
@@ -3,6 +3,63 @@ import _capitalize from "lodash/capitalize";
 import fs from "fs-extra";
 import { migrateUIFolder } from "../migrateUIFolder";
 import path from "path";
+import { ContentRecord, DataModel } from "@activeviam/activeui-sdk";
+import { ErrorReport, OutcomeCounters } from "../migration.types";
+import _fromPairs from "lodash/fromPairs";
+
+type MigrationFunction = (
+  legacyUIFolder: ContentRecord,
+  counters: OutcomeCounters,
+  errorReport: ErrorReport,
+  {
+    servers,
+    keysOfWidgetPluginsToRemove,
+    legacyPivotFolder,
+    doesReportIncludeStacks,
+  }: {
+    servers: {
+      [serverKey: string]: {
+        dataModel: DataModel;
+        url: string;
+      };
+    };
+    keysOfWidgetPluginsToRemove?: string[] | undefined;
+    doesReportIncludeStacks: boolean;
+    legacyPivotFolder?: ContentRecord<any> | undefined;
+  },
+) => Promise<[ContentRecord, OutcomeCounters, ErrorReport?]>;
+
+const initialCounters = _fromPairs(
+  ["dashboards", "widgets", "filters", "folders"].map((type) => [
+    type,
+    {
+      success: 0,
+      partial: 0,
+      failed: 0,
+      removed: 0,
+    },
+  ]),
+  // _fromPairs returns a Dictionary.
+  // In this case, the keys used correspond to the attributes of OutcomeCounters.
+) as OutcomeCounters;
+
+const compose =
+  (...migrationFunctions: MigrationFunction[]) =>
+  async (
+    ...contentToMigrate: Parameters<MigrationFunction>
+  ): ReturnType<MigrationFunction> => {
+    if (migrationFunctions.length === 0) {
+      return [contentToMigrate[0], initialCounters, {}];
+    }
+    const [migratedUIFolder, counters, errorReport] =
+      await migrationFunctions[0](...contentToMigrate);
+    return compose(...migrationFunctions.slice(1))(
```

Comment:
No need to slice, you can destructure instead which is more efficient: `[firstFunction, ...rest]`.

But it seems to me that an implementation with `reduce` or even a `for` loop would be shorter ?

Reply:
Related to what's coming up in [this PR](https://github.com/activeviam/activeui-migration/pull/55), I changed the signature of `migrate_43_to_50` and used destructuration, which makes the `compose` simpler. Let me know what you think.